trail/upload/process_uploads/standardized_dataclasses.py

```diff
@@ -0,0 +1,208 @@
+
+__all__ = ["StandardizedWcsKeys",
+           "StandardizedMetadataKeys",
+           "StandardizedHeaderKeys",]
+
+
+dtype = np.dtype([("names", str, 30), ("types", object),  ("defaults", object),
+                  ("required", bool),("closeEquality", bool)])
+
+wcsAttributes = np.array([
+    ("wcs_radius",   float, None, True, True),
+    ("wcs_center_x", float, None, True, True),
+    ("wcs_center_y", float, None, True, True),
+    ("wcs_center_z", float, None, True, True),
+    ("wcs_corner_x", float, None, True, True),
+    ("wcs_corner_y", float, None, True, True),
+    ("wcs_corner_z", float, None, True, True)],
+                         dtype=dtype)
+
+metaAttrs = np.array([
+    ("obs_lon",           float, None, True,  True),
+    ("obs_lat",           float, None, True,  True),
+    ("obs_height",        float, None, True,  True),
+    ("telescope",          str, None, True,  False),
+    ("datetime_begin",     str, None, True,  False),
+    ("datetime_end",       str, None, True,  False),
+    ("processor_name",     str, None, False, False),
+    ("standardizer_name",  str, None, False, False),
+    ("instrument",         str, None, False, False),
+    ("science_program",    str, None, False, False),
+    ("exposure_duration",  str, None, False, False),
+    ("physical_filter",    str, None, False, False)],
+                     dtype=dtype)
+
+
+def get_dict_subset(data, keys):
+    return {key: data.get(key, None) for key in keys}
+
+
+class StandardizedKeysMixin:
+    """Mix-in class for StandardizeKeys classes."""
+
+    @classmethod
+    def fromDictSubset(cls, data):
+        return cls(**get_dict_subset(data, cls.keys))
+
+    def values(self):
+        return [getattr(self, key) for key in self.keys]
+
+    def toDict(self):
+        # asdict use here would be nice, but is much slower...
+        return {key: getattr(self, key) for key in self.keys}
+
+    def valid(self):
+        for key, keytype in zip(self._requiredKeys, self.keyTypes):
+            if not (getattr(self, key) and type(getattr(self, key)) == keytype):
+                return False
+        return True
+
+
+@dataclass(order=True)
+class StandardizedWcsBase(ABC, StandardizedKeysMixin):
+    keys: ClassVar[list] = wcsAttributes["names"]
+    keyTypes: ClassVar[list] = wcsAttributes["types"]
+    _requiredKeys: ClassVar[list] = keys
+    _closeEqualityKeys = keys
+
+    def isCloseTo(self, other, **kwargs):
+        return np.allclose(self.values(), other.values(), **kwargs)
+
+
+@dataclass(order=True)
+class StandardizedMetadataBase(ABC, StandardizedKeysMixin):
+    keys: ClassVar[list] = metaAttrs["names"]
+    keyTypes: ClassVar[list] = metaAttrs["types"]
+    _closeEqualityKeys = metaAttrs[metaAttrs["closeEquality"] == True]["names"]
+
+    _requiredKeys: ClassVar[list] = metaAttrs[metaAttrs["required"] == True]["names"]
+    _optionalKeys: ClassVar[list] = metaAttrs[metaAttrs["required"] == False]["names"]
+
+    def isCloseTo(self, other, **kwargs):
```

Comment:
Docstring plz, in fact the whole Mixin class could use some more docstrings

Reply:
This was merged with models and file trashed, models now have docstrings.